batch2/batch/driver/scheduler.py

```diff
@@ -19,6 +20,59 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([worker.free_cores_mcpu for worker in self.inst_pool.healthy_instances_by_free_cores])
+
+        sorted_running_cores_mcpu = sortedcontainers.SortedSet(key=lambda x: x[1])
+        sorted_total_cores_mcpu = sortedcontainers.SortedSet(key=lambda x: x[1])
+
+        running_cores_mcpu = {}
+        allocated_cores = {}
+
+        records = self.db.execute_and_fetchall(
+            '''
+SELECT user, ready_cores_mcpu, running_cores_mcpu
+FROM user_resources
+WHERE ready_cores_mcpu > 0;
+''')
+
+        async for record in records:
+            running_cores_mcpu[record['user']] = record['running_cores_mcpu']
+            sorted_running_cores_mcpu.add(
+                (record['user'], record['running_cores_mcpu']))
+            sorted_total_cores_mcpu.add(
+                (record['user'], record['running_cores_mcpu'] + record['ready_cores_mcpu']))
```

Comment:
Following the boat analog, I think what you want is:

 - allocating_users_by_total_cores which starts empty (no boats on the water in the beginning), and
 - pending_users_by_running_cores, which starts with all users (the boat's original height)